amazon-kinesis-client/src/test/java/software/amazon/kinesis/coordinator/SchedulerTest.java

```diff
@@ -784,6 +777,55 @@ private final void testMultiStreamNewStreamsAreSyncedAndStaleStreamsAreNotDelete
                 scheduler.staleStreamDeletionMap().keySet());
     }
 
+
+    @Test
+    public void testKinesisStaleDeletedStreamCleanup() throws ProvisionedThroughputException, InvalidStateException, DependencyException {
+        List<StreamConfig> streamConfigList1 = createDummyStreamConfigList(1,6);
+        List<StreamConfig> streamConfigList2 = createDummyStreamConfigList(1,4);
+        when(multiStreamTracker.streamConfigList()).thenReturn(streamConfigList1, streamConfigList2);
+        when(multiStreamTracker.formerStreamsLeasesDeletionStrategy()).thenReturn(new AutoDetectionAndDeferredDeletionStrategy() {
+            @Override public Duration waitPeriodToDeleteFormerStreams() {
+                return Duration.ofDays(1);
+            }
+        });
+        retrievalConfig = new RetrievalConfig(kinesisClient, multiStreamTracker, applicationName)
+                .retrievalFactory(retrievalFactory);
+        scheduler = spy(new Scheduler(checkpointConfig, coordinatorConfig, leaseManagementConfig, lifecycleConfig,
+                metricsConfig, processorConfig, retrievalConfig));
+        when(scheduler.shouldSyncStreamsNow()).thenReturn(true);
+
+        // when KCL starts it starts with tracking 5 stream
+        assertEquals(5, scheduler.currentStreamConfigMap().size());
+        assertEquals(0, scheduler.staleStreamDeletionMap().size());
+
+        // 2 Streams are no longer needed to be consumed
+        Set<StreamIdentifier> syncedStreams2 = scheduler.checkAndSyncStreamShardsAndLeases();
+        assertEquals(5, scheduler.currentStreamConfigMap().size());
+        assertEquals(2, scheduler.staleStreamDeletionMap().size());
+
+        // One the stream is deleted from Kinesis side
+        scheduler.deletedStreamListProvider().add(createDummyStreamConfig(5).streamIdentifier());
+
+        Set<StreamIdentifier> syncedStreams3 = scheduler.checkAndSyncStreamShardsAndLeases();
+        //assert kinesis deleted stream is cleaned up from KCL in memory state.
+        assertEquals(4, scheduler.currentStreamConfigMap().size());
+        assertEquals(1, scheduler.staleStreamDeletionMap().size());
+
```

Comment:
Could we verify the remaining streams are ones that weren't deleted? We should also try deletion one more time to verify the `deletedStreamListProvider` behaves as expected by reseting the deletion set after purged. (It can probably be a new unit test class)

Reply:
added assert to verify that remaining streams are the ones that werent deleted.

Added assert to same unit tests which purges the provider again and verify that this time size is 0 (expected)